tss-esapi/src/abstraction/transient/key_attestation.rs

```diff
@@ -0,0 +1,203 @@
+    /// the credential
+    ///
+    /// **Note**: If no `key` is given, the default Endorsement Key
+    /// will be used.  
+    pub fn get_make_cred_params(
+        &mut self,
+        object: ObjectWrapper,
+        key: Option<ObjectWrapper>,
+    ) -> Result<MakeCredParams> {
+        let object_handle = self.load_key(object.params, object.material, None)?;
+        let (object_public, object_name, _) =
+            self.context.read_public(object_handle).or_else(|e| {
+                self.context.flush_context(object_handle.into())?;
+                Err(e)
+            })?;
+        self.context.flush_context(object_handle.into())?;
+
+        let public = TPM2B_PUBLIC::from(object_public);
+        let public = unsafe {
+            std::mem::transmute::<TPMT_PUBLIC, [u8; std::mem::size_of::<TPMT_PUBLIC>()]>(
+                public.publicArea,
+            )
+        };
+        let attesting_key_pub = match key {
+            None => {
+                let key_handle =
+                    ek::create_ek_object(&mut self.context, AsymmetricAlgorithm::Rsa, None)?;
+                let (attesting_key_pub, _, _) =
+                    self.context.read_public(key_handle).or_else(|e| {
+                        self.context.flush_context(key_handle.into())?;
+                        Err(e)
+                    })?;
+                self.context.flush_context(key_handle.into())?;
+
+                attesting_key_pub.try_into()?
+            }
+            Some(key) => key.material.public,
+        };
+        Ok(MakeCredParams {
+            name: object_name.value().to_vec(),
+            public: public.to_vec(),
+            attesting_key_pub,
+        })
+    }
+
+    /// Perform an ActivateCredential operation for the given object
+    ///
+    /// # Parameters
+    ///
+    /// * `object` - the object whose TPM name is included in the credential
+    /// * `key` - the key used to encrypt the secret that wraps the credential
+    /// * `credential_blob` - encrypted credential that will be returned by the
+    /// TPM
+    /// * `secret` - encrypted secret that was used to encrypt the credential
+    ///
+    /// **Note**: if no `key` is given, the default Endorsement Key
+    /// will be used. You can find more information about the default Endorsement
+    /// Key in the [ek] module.
+    pub fn activate_credential(
+        &mut self,
+        object: ObjectWrapper,
+        key: Option<ObjectWrapper>,
+        credential_blob: Vec<u8>,
+        secret: Vec<u8>,
+    ) -> Result<Vec<u8>> {
+        let credential_blob = IDObject::try_from(credential_blob)?;
+        let secret = EncryptedSecret::try_from(secret)?;
+        let object_handle = self.load_key(object.params, object.material, object.auth)?;
+        let session_2;
+        let key_handle = match key {
+            None => {
+                // No key was given, use the EK. This requires using a Policy session
+                session_2 = self
```

Comment:
I think it should probably put the two cases in separate private methods so its get easier to read.

Reply:
Ok, done, though I kept the `match key` - I find it more intuitive to follow